cql3/statements/call_statement.cc

```diff
@@ -0,0 +1,141 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "cql3/statements/call_statement.hh"
+#include "cql3/statements/raw/call_statement.hh"
+#include "db/config.hh"
+#include "database.hh"
+#include "transport/messages/result_message.hh"
+#include "db/builtin_routine.hh"
+
+namespace cql3::statements {
+
+call_statement::call_statement(::shared_ptr<cf_name> cf_name,
+            seastar::shared_ptr<db::builtin_routine> proc,
+            uint32_t bound_terms,
+            std::unordered_map<sstring, sstring> params,
+            cql_stats& stats)
+    // todo: read timeout, write timeout, or ddl tiemout,
+    // depending on call body
+    : cql_statement_opt_metadata(&timeout_config::write_timeout)
+    , _cf_name(std::move(cf_name))
+    , _proc(proc)
+    , _bound_terms{bound_terms}
+    , _params(std::move(params))
+    , _stats(stats)
+    // todo: correct ks_selector for user-defined procedures
+    , _ks_sel(ks_selector::SYSTEM) {
+
+    if (_proc->has_output()) {
+        std::vector<lw_shared_ptr<column_specification>> columns;
+        for (auto&& col : _proc->metadata) {
+            // Add the only supported result column to result set metadata
+            auto result_column = make_lw_shared<cql3::column_specification>(_cf_name->get_keyspace(),
+                _cf_name->get_column_family(), make_shared<cql3::column_identifier>(col.first, false),
+                col.second);
+
+            columns.push_back(std::move(result_column));
+        }
+        _metadata = seastar::make_shared<cql3::metadata>(std::move(columns));
+    }
+}
+
+uint32_t call_statement::get_bound_terms() const {
+    return _bound_terms;
+}
+
+future<> call_statement::check_access(service::storage_proxy& proxy, const service::client_state& state) const {
+    // todo: check auth::CALL permission
+    return make_ready_future<>();
+}
+
+void
+call_statement::validate(service::storage_proxy&, const service::client_state& state) const {
+    // todo
+}
+
+bool call_statement::depends_on_column_family(const sstring& cf_name) const {
+    return false;
+}
+
+bool call_statement::depends_on_keyspace(const sstring& ks_name) const {
+    // todo : return true if ks_name = procedure keyspace
+    return false;
+}
+
+future<::shared_ptr<cql_transport::messages::result_message>>
+call_statement::execute(query_processor& qp, service::query_state& qs, const query_options& options) const {
+
+    inc_cql_stats(qs.get_client_state().is_internal());
+
+    db::builtin_routine::parameter_map params(_params.begin(), _params.end());
+
+    return _proc->call(std::move(params)).then([this](db::builtin_routine::output output) {
+        using rm = cql_transport::messages::result_message;
+        using srm = ::shared_ptr<rm>;
+        if (_proc->has_output()) {
+            auto result_set = std::make_unique<cql3::result_set>(_metadata);
+            for (auto &&row : output) {
```

Comment:
Some heavy lifting is done inside the loop - perhaps it makes sense to use seastar's `for_each`

Reply:
(or coroutines :innocent: , in which case each iteration can call `co_await coroutine::maybe_yield()`)

Reply:
(that comment still applies)